GSoC Simultaneous Trajectory Execution (#769)
moveit/moveit.ros.org: _posts/2023-01-12-gsoc-simultaneous-trajectory-execution.md

---
author: Cristian C. Beltran-Hernandez
comments: false
date: 2023-01-12
layout: post
title: Simultaneous Trajectory Execution
media_type: image
media_link: /assets/images/blog_posts/gsoc-2022-badge.svg
description: Google Summer of Code 2022 project's result
categories:
- MoveIt
- Google
---

> This post summarizes my contributions during the [Google Summer of Code 2022](https://summerofcode.withgoogle.com/programs/2022/projects/mpusZVc2). I worked on extending the trajectory execution capabilities of MoveIt. 

Reviewer: ```suggestion
This post summarizes my contributions during the [Google Summer of Code 2022](https://summerofcode.withgoogle.com/programs/2022/projects/mpusZVc2) where I have extended the trajectory execution capabilities of MoveIt. 
```

# Motivation

Before this project, MoveIt only allowed the execution of trajectories in strict sequential order, which limited the options to simultaneously control two or more robots to only synchronized motions.

The goal was then to allow the simultaneous execution of multiple trajectories on systems with multiple robots.  For example, in a dual-arm robotic system, each arm can execute a different set of trajectories without needing to wait for the other arm to finish moving (sequential execution) or manually synchronizing the motion of both arms into a single trajectory. To preserve collision-free guarantees an extra collision check needs to be performed right before the execution of new trajectories to prevent collisions with active trajectories. 

> ## Example use cases:

Reviewer: ```suggestion
## Example Use Cases
```

1. Several trajectories are planned and executed through the `MoveIt Motion Planning Rviz` plugin. 

> ![simultaneous-execution-rviz](https://user-images.githubusercontent.com/3798796/196121126-6842200c-6a81-42cc-bc7a-b9f7200d7a2b.gif)

Reviewer 2: Shouldn't the GIFs be added to the web site directly? It seems like content under https://user-images.githubusercontent.com/ could disappear in the future.

2. Trajectories being planned and executed from different scripts (Rviz + Python script)
![simultaneous-execution-python-rviz](https://user-images.githubusercontent.com/3798796/196125955-68b9698a-45ce-4a3c-9bac-668b7d4f15d6.gif)
3. **Real application** examples can be found [here](https://github.com/ros-planning/moveit/pull/2810) such as executing screwing task independently with each arm.
![simultaneous-screwing](https://user-images.githubusercontent.com/3798796/212067483-b8ff2c2f-994e-42bd-a411-66a27e91eeb1.gif)

> ## Feature description

Reviewer: ```suggestion
## Feature Description
```

- Event-based execution system: Events related to the execution of trajectories are pushed to a queue where they are processed sequentially.

>   When a new trajectory is pushed, it is immediately validated, by checking that the required controllers are active and not in use and that there are no collisions with active trajectories or the current state of the planning scene. Invalid trajectories are rejected. Valid trajectories are sent for execution to the corresponding controllers. 

Reviewer: ```suggestion
  When a new trajectory is pushed, it is immediately validated by checking that all required controllers are active and available, and that there is no collision with any active trajectory or the current state. Invalid trajectories are rejected. Valid trajectories are sent for execution to the corresponding controllers. 
```
